### ISSUE TITLE
Fix semantics of variable domains

### ISSUE THREAD

**Issue by ghackebeil** [Pyomo/pyomo#86]
One use case that we would want to be trivial is relaxing a discrete variable without changing its bounds. The naive approach would be something like the following

```python
model.y = Var(domain=Binary)
print(model.y.bounds) # -> (0, 1)                                                            
model.y.domain = Reals
print(model.y.bounds) # -> (None, None)
```

I am proposing the following approach: Extract the bounds from the domain and store the domain type on the variable. There would no longer be a domain attribute that one could read (but it could be set). Also allow changing the domain type.

```python
x = variable(domain=Binary)
assert x.domain_type = IntegerSet
assert x.lb = 0
assert x.ub = 1

# relax integrality without affecting the bounds
x.domain_type = RealSet
assert x.domain_type = RealSet
assert x.lb = 0
assert x.ub = 1

# assigning to the domain property setter always overwrites the bounds
# (but there is no domain property to read from)
x.domain = Integers
assert x.domain_type = IntegerSet
assert x.lb = None
assert x.ub = None

# rule: you can not mix a bound declaration with
#       a domain declaration in the constructor
#       if the domain bound is finite
variable(domain=Reals, lb=-1, ub=10)     # OK
variable(domain=NonNegativeReals, ub=1)  # OK
variable(domain=NonNegativeReals, lb=1)  # ERROR
```

The point of the rule about restricting certain domain and bound declarations in the constructor is about avoiding having to evaluate the lb and ub assignments (or compare them to anything other than 'None'), which allows us to natively support mutable expressions as bounds.

All of this is currently prototyped on the component_storage_interface branch. Examples can be found here: https://github.com/Pyomo/pyomo/tree/component_storage_interface/examples/kernel

**Comment by qtothec** [comment]
I believe this behavior was addressed at some point?

**Comment by ghackebeil** [comment]
I've implemented this in the kernel `variable` object. The issue persists on `Var`.

I've never heard any feedback on the issue or whether the fix I've suggested should be moved into `Var`.

**Comment by qtothec** [comment]
Yup, looks like the issue persists on `Var`. I agree that this is undesirable behavior.

Another potential idiom to go with is to warn when the user is implicitly relaxing the bounds of a variable and push them to use a transformation instead,
